bindings/java/src/org/sleuthkit/datamodel/AbstractFile.java

```diff
@@ -499,8 +510,75 @@ public void setSha256Hash(String sha256Hash) {
 	 */
 	public String getSha256Hash() {
 		return this.sha256Hash;
+	}	
+	
+	/**
+	 * Gets the attributes of this File
+	 * @return
+	 * @throws TskCoreException 
+	 */
+	public List<Attribute> getAttributes() throws TskCoreException {
+		if (!loadedAttributesCacheFromDb.get()) {
+			ArrayList<Attribute> attributes = getSleuthkitCase().getFileAttributes(this);
+			fileAttributesCache.clear();
+			fileAttributesCache.addAll(attributes);
+			loadedAttributesCacheFromDb.set(true);
+		}
+		return Collections.unmodifiableList(fileAttributesCache);
 	}
 
+	/**
+	 * Adds a collection of attributes to this file in a single operation 
+	 * within a transaction supplied by the caller.
+	 *
+	 * @param attributes        The collection of attributes.
+	 * @param caseDbTransaction The transaction in the scope of which the
+	 *                          operation is to be performed, managed by the
+	 *                          caller. if Null is passed in a local transaction
+	 *							will be created and used. 
+	 *
+	 * @throws TskCoreException         If an error occurs and the attributes
+	 *                                  were not added to the artifact.
+	 * @throws IllegalArgumentException If <code>attributes</code> is
+	 *                                  null or empty.
+	 */
+	public void addAttributes(Collection<Attribute> attributes, final SleuthkitCase.CaseDbTransaction caseDbTransaction) throws TskCoreException {
+
+		if (Objects.isNull(attributes) || attributes.isEmpty()) {
+			throw new IllegalArgumentException("null or empty attributes passed to addAttributes");
+		}
+		boolean isLocalTransaction = Objects.isNull(caseDbTransaction);
+		SleuthkitCase.CaseDbTransaction localTransaction = isLocalTransaction ? getSleuthkitCase().beginTransaction() : null;		
+		SleuthkitCase.CaseDbConnection connection = isLocalTransaction ? localTransaction.getConnection() : caseDbTransaction.getConnection();
+		
+		try {
+			for (final Attribute attribute : attributes) {
+				attribute.setAttributeOwnerId(getId()); 
+				attribute.setCaseDatabase(getSleuthkitCase());
+				getSleuthkitCase().addFileAttribute(attribute, connection);
+			}
+			
+			if(isLocalTransaction) {
+				localTransaction.commit();
+				localTransaction = null;
+			}
+			// append the new attributes if cache is already loaded. 
+			if (loadedAttributesCacheFromDb.get()) {
+				fileAttributesCache.addAll(attributes);
+			}
+		} catch (SQLException ex) {
+			throw new TskCoreException("Error adding file attributes", ex);
+		} finally {
+			if (isLocalTransaction && null != localTransaction) {
+				try {
+					localTransaction.rollback();
+				} catch (TskCoreException ex2) {
+					LOGGER.log(Level.SEVERE, "Failed to rollback transaction after exception", ex2);
+				}
+			}
+		}
```

Comment:
The logic for committing or rolling back the transaction is kind of complex here. I think it can be simplified. How about putting the commit in the try block and the rollback in the catch block and eliminating the finally block?